tls/s2n_x509_validator.c

```diff
@@ -549,3 +554,66 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(
     return ret_val;
 #endif /* S2N_OCSP_STAPLING_SUPPORTED */
 }
+
+S2N_RESULT s2n_validate_certificate_signature(struct s2n_connection *conn, X509 *x509_cert)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(x509_cert);
+
+    const struct s2n_security_policy *security_policy;
+    GUARD_AS_RESULT(s2n_connection_get_security_policy(conn, &security_policy));
+
+    if (security_policy->certificate_signature_preferences == NULL) {
+        return S2N_RESULT_OK;
+    }
+
+    X509_NAME *issuer_name = X509_get_issuer_name(x509_cert);
+    ENSURE_REF(issuer_name);
+
+    X509_NAME *subject_name = X509_get_subject_name(x509_cert);
+    ENSURE_REF(subject_name);
+
+    /* Do not validate any self-signed certificates */
+    if (X509_NAME_cmp(issuer_name, subject_name) == 0) {
+        return S2N_RESULT_OK;
+    }
+
+    bool out = false;
+    GUARD_RESULT(s2n_is_certificate_sig_scheme_supported(conn, x509_cert, security_policy->certificate_signature_preferences, &out));
+    ENSURE(out == true, S2N_ERR_CERT_UNTRUSTED);
```

Comment:
Since you immediately translate `out` to an error, is there any benefit in having a separate true/false result? Maybe you should rename `s2n_is_certificate_sig_scheme_supported` to `s2n_validate_sig_scheme_supported` and just error if the scheme is not supported.